File: php/src/Models/ApplicationOut.php
<?php

// this file is @generated
declare(strict_types=1);

namespace Svix\Models;

class ApplicationOut implements \JsonSerializable
{
    private array $setFields = [];

    /**
     * @param string|null $uid          optional unique identifier for the application
     * @param string      $name         application name for human consumption
     * @param int|null    $throttleRate Maximum messages per second to send to this application.
     *
     * Outgoing messages will be throttled to this rate.
     * @param string                $id       the Application's ID
     * @param array<string, string> $metadata
     */
    private function __construct(
        public readonly string $name,
        public readonly string $id,
        public readonly \DateTimeImmutable $createdAt,
        public readonly \DateTimeImmutable $updatedAt,
        public readonly array $metadata,
        public readonly ?string $uid = null,
        public readonly ?int $throttleRate = null,
        array $setFields = [],
    ) {
        $this->setFields = $setFields;
    }

    /**
     * Create an instance of ApplicationOut with required fields.
     */
    public static function create(
        string $name,
        string $id,
        \DateTimeImmutable $createdAt,
        \DateTimeImmutable $updatedAt,
        array $metadata,
    ): self {
        return new self(
            uid: null,
            name: $name,
            throttleRate: null,
            id: $id,
            createdAt: $createdAt,
            updatedAt: $updatedAt,
            metadata: $metadata,
            setFields: ['name' => true, 'id' => true, 'createdAt' => true, 'updatedAt' => true, 'metadata' => true]
        );
    }

    public function withUid(?string $uid): self
    {
        $setFields = $this->setFields;
        $setFields['uid'] = true;

        return new self(
            uid: $uid,
            name: $this->name,
            throttleRate: $this->throttleRate,
            id: $this->id,
            createdAt: $this->createdAt,
            updatedAt: $this->updatedAt,
            metadata: $this->metadata,
            setFields: $setFields
        );
    }

    public function withThrottleRate(?int $throttleRate): self
    {
        $setFields = $this->setFields;
        $setFields['throttleRate'] = true;

        return new self(
            uid: $this->uid,
            name: $this->name,
            throttleRate: $throttleRate,
            id: $this->id,
            createdAt: $this->createdAt,
            updatedAt: $this->updatedAt,
            metadata: $this->metadata,
            setFields: $setFields
        );
    }

    public function jsonSerialize(): mixed
    {
        $data = [
            'name' => $this->name,
            'id' => $this->id,
            'createdAt' => $this->createdAt->format('c'),
            'updatedAt' => $this->updatedAt->format('c'),
            'metadata' => $this->metadata];

        if (isset($this->setFields['uid'])) {
            $data['uid'] = $this->uid;
        }
        if (isset($this->setFields['throttleRate'])) {
            $data['throttleRate'] = $this->throttleRate;
        }

        return \Svix\Utils::newStdClassIfArrayIsEmpty($data);
    }

    /**
     * Create an instance from a mixed obj.
     */
    public static function fromMixed(mixed $data): self
    {
        return new self(
            uid: \Svix\Utils::deserializeString($data, 'uid', false, 'ApplicationOut'),
            name: \Svix\Utils::deserializeString($data, 'name', true, 'ApplicationOut'),
            throttleRate: \Svix\Utils::deserializeInt($data, 'throttleRate', false, 'ApplicationOut'),
            id: \Svix\Utils::deserializeString($data, 'id', true, 'ApplicationOut'),
            createdAt: \Svix\Utils::deserializeDt($data, 'createdAt', true, 'ApplicationOut'),
            updatedAt: \Svix\Utils::deserializeDt($data, 'updatedAt', true, 'ApplicationOut'),
            metadata: \Svix\Utils::getValFromJson($data, 'metadata', true, 'ApplicationOut')
        );
    }

    /**
     * Create an instance from a json string.
     */
    public static function fromJson(string $json): self
    {
        $data = json_decode(json: $json, associative: true, depth: 512, flags: JSON_THROW_ON_ERROR);

        return self::fromMixed($data);
    }
}

File: java/lib/src/main/java/com/svix/models/ApplicationOut.java
// This file is @generated
package com.svix.models;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.svix.Utils;

import lombok.EqualsAndHashCode;
import lombok.ToString;

import java.time.OffsetDateTime;
import java.util.HashMap;
import java.util.Map;

@ToString
@EqualsAndHashCode
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonAutoDetect(getterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
public class ApplicationOut {
    @JsonProperty private OffsetDateTime createdAt;
    @JsonProperty private String id;
    @JsonProperty private Map<String, String> metadata;
    @JsonProperty private String name;
    @JsonProperty private Long rateLimit;
    @JsonProperty private Long throttleRate;
    @JsonProperty private String uid;
    @JsonProperty private OffsetDateTime updatedAt;

    public ApplicationOut() {}

    public ApplicationOut createdAt(OffsetDateTime createdAt) {
        this.createdAt = createdAt;
        return this;
    }

    /**
     * Get createdAt
     *
     * @return createdAt
     */
    @javax.annotation.Nonnull
    public OffsetDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(OffsetDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public ApplicationOut id(String id) {
        this.id = id;
        return this;
    }

    /**
     * The Application's ID.
     *
     * @return id
     */
    @javax.annotation.Nonnull
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public ApplicationOut metadata(Map<String, String> metadata) {
        this.metadata = metadata;
        return this;
    }

    public ApplicationOut putMetadataItem(String key, String metadataItem) {
        if (this.metadata == null) {
            this.metadata = new HashMap<>();
        }
        this.metadata.put(key, metadataItem);

        return this;
    }

    /**
     * Get metadata
     *
     * @return metadata
     */
    @javax.annotation.Nonnull
    public Map<String, String> getMetadata() {
        return metadata;
    }

    public void setMetadata(Map<String, String> metadata) {
        this.metadata = metadata;
    }

    public ApplicationOut name(String name) {
        this.name = name;
        return this;
    }

    /**
     * Application name for human consumption.
     *
     * @return name
     */
    @javax.annotation.Nonnull
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    @Deprecated
    public ApplicationOut rateLimit(Long rateLimit) {
        this.rateLimit = rateLimit;
        return this;
    }

    /**
     * Deprecated, use `throttleRate` instead.
     *
     * @return rateLimit
     */
    @javax.annotation.Nullable
    @Deprecated
    public Long getRateLimit() {
        return rateLimit;
    }

    @Deprecated
    public void setRateLimit(Long rateLimit) {
        this.rateLimit = rateLimit;
    }

    public ApplicationOut throttleRate(Long throttleRate) {
        this.throttleRate = throttleRate;
        return this;
    }

    /**
     * Maximum messages per second to send to this application.
     *
     * <p>Outgoing messages will be throttled to this rate.
     *
     * @return throttleRate
     */
    @javax.annotation.Nullable
    public Long getThrottleRate() {
        return throttleRate;
    }

    public void setThrottleRate(Long throttleRate) {
        this.throttleRate = throttleRate;
    }

    public ApplicationOut uid(String uid) {
        this.uid = uid;
        return this;
    }

    /**
     * Optional unique identifier for the application.
     *
     * @return uid
     */
    @javax.annotation.Nullable
    public String getUid() {
        return uid;
    }

    public void setUid(String uid) {
        this.uid = uid;
    }

    public ApplicationOut updatedAt(OffsetDateTime updatedAt) {
        this.updatedAt = updatedAt;
        return this;
    }

    /**
     * Get updatedAt
     *
     * @return updatedAt
     */
    @javax.annotation.Nonnull
    public OffsetDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(OffsetDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    /**
     * Create an instance of ApplicationOut given an JSON string
     *
     * @param jsonString JSON string
     * @return An instance of ApplicationOut
     * @throws JsonProcessingException if the JSON string is invalid with respect to ApplicationOut
     */
    public static ApplicationOut fromJson(String jsonString) throws JsonProcessingException {
        return Utils.getObjectMapper().readValue(jsonString, ApplicationOut.class);
    }

    /**
     * Convert an instance of ApplicationOut to an JSON string
     *
     * @return JSON string
     */
    public String toJson() throws JsonProcessingException {
        return Utils.getObjectMapper().writeValueAsString(this);
    }
}
